Prevent future ADK/SDF scan-register problems.

steveri opened the pull request:
A recent change to the TSMC16 adk repo re-introduced an old problem, when synthesis started using scan-enabled flipflops in place of mux-register combos; see issue https://github.com/StanfordAHA/garnet/issues/833 for the gory details. Long story short, this causes multiple problems and prevents successful chip builds.

And so, since this has happened at least twice now, I added some code to help detect when/if it happens again in the future. You'll see these changes in the `construct.py` files for `Tile_PE` and `Tile_MemCore`. Also, I added an extra check to the CI setup code `setup-buildkite.sh` to verify that the tsmc16 repo is up-to-date before embarking on a long build.

We also had some recent docker problems that required CI debugging, during which I found errors and inadequacies in the code for building with docker images other than the latest version of "latest". The changes in `gen_rtl.sh` are designed to fix these.

I am also taking this opportunity to make some tweaks in the CI toolkit, particularly in terms of better build names for `mem_tile_only` and `pe_tile_only` pipelines (occasionally used for debugging the flow). There are two new pipelines in the tool kit, `pe_tile_w_steps` and `pe_tile_w_cached_steps`, both of which allow running specific steps in subgraphs, with and without cached collateral, along with tweaks to a helper script `get-step-context.sh`.

So. Summary of file changes in this pull:

* FAIL if bad SDFQ regs found in pe or mem tile after synthesis
```
mflowgen/Tile_MemCore/construct.py
mflowgen/Tile_PE/construct.py
```
* FAIL if tsmc repo not current
```
mflowgen/bin/setup-buildkite.sh
```
* Mechanism for using docker images other than most-recent version of "latest" image. For now, you have to modify the source code to change the default docker image; a future tweak would maybe make this into an env var parameter.
```
common/rtl/gen_rtl.sh
```
* Tools for debugging buildkite full-chip build
```
pipelines/mem_tile_only.yml
pipelines/pe_tile_only.yml
pipelines/pe_tile_w_cached_steps.yml
pipelines/pe_tile_w_steps.yml
mflowgen/bin/get-step-context.sh
```

steveri commented:
Also: this branch has successfully completed a full 36-hour chip build, see https://buildkite.com/tapeout-aha/fullchip/builds/363

alexcarsello commented on mflowgen/common/rtl/gen_rtl.sh:
```diff
@@ -46,29 +48,39 @@ else
 
       # See common/rtl/configure.yml for "which_image" setting; default = "latest"
-      if [ "$which_image" == "" ]; then which_image=latest; fi
+      if [ "$which_image" == "" ]; then which_image=":latest"; fi
```
Also, if you add `which_image` as a node param to the configure.yml, this line is no longer necessary.

steveri replied:
I ran into trouble in the case where, if you have a named docker image, you can do e.g. 
```
docker pull stanfordaha/garnet:$name
```
 but if you just have a sha it now has to be
```
docker pull stanfordaha/garnet@$name
```
I originally had code that detected whether $which_image was name or sha and manipulated it accordingly, but it was really ugly so I opted for this (simpler? but maybe uglier?) approach.

alexcarsello replied:
Ok that makes sense

alexcarsello replied:
Just make a comment in the configure.yml explaining that requirement.

steveri replied:
See prev comment.

alexcarsello reviewed (changes requested):
See the inline comments

steveri commented on mflowgen/Tile_MemCore/construct.py:
```diff
@@ -366,6 +366,10 @@ def construct():
+  # The presence of SDF registers after synthesis means adk got corrupted somehow.
+  # The registers cause short circuits that are not found until much later, after LVS.
+  # This has happened several times now, see .e.g. https://buildkite.com/tapeout-aha/fullchip/builds/358
+  synth.extend_postconditions( ["assert 'SDFQ' not in File( 'outputs/design.v' )"] )
```
I agree that the test is technology-specific, but then of course `construct.py` is technology-specific, see above where it says `adk_name = 'tsmc16'` :)

But I do see your point. I could change the test to something like
```
if (adk_name == "tsmc16"); then extend_postconditions ...
```
and/or if you guarantee that the problem will never happen again, I can take out the test(s) entirely...otherwise I see no harm in keeping them...

alexcarsello replied:
There's one line that selects the ADK, but ideally we keep the rest of the code technology agnostic here. I'd rather remove it entirely.

The other issue here is that this check is only being added to the mem and PE flows, but of course, it's not mem or PE specific. Ideally, if we wanted to add this check, we'd add it across all of our flows. But the way mflowgen is right now makes that tedious. Some changes are coming that should hopefully help with that. Regardless, I don't think we need this check anyway.

steveri replied:
Removed.

steveri commented on mflowgen/bin/setup-buildkite.sh:
```diff
@@ -426,7 +426,17 @@ if [ "$USER" == "buildkite-agent" ]; then
         pushd $d
             git branch -v
-            ba=`git branch -v | awk '{print $4}'` # E.g. '[ahead' or '[behind'
+            # Check that tsmc16 is pointing to master branch.
+            ba=`git branch -v | awk '{print $2; exit}'` 
+            if [ "$ba" != "master" ]; then
+                echo "---------------------------------------------------------"
+                echo "**ERROR looks like tsmc16 repo is not on master branch."
+                echo "---------------------------------------------------------"
+                return 13 || exit 13
+            fi
+
+            # Checks to see that adk repo is not ahead or behind origin
+            ba=`git branch -v | awk '{print $4; exit}'` # E.g. '[ahead' or '[behind'
```
Good point. git clone is a better fix. I'll see if I can get that working...not sure atm what kind of permissions buildkite-agent has or needs wrt the tsmc16 repo

steveri commented on mflowgen/common/rtl/gen_rtl.sh:
```diff
@@ -46,29 +48,39 @@ else
       # new code (below) allows use of any image based on new "which_image" parm
 
       # See common/rtl/configure.yml for "which_image" setting; default = "latest"
```
Great idea, I will do that.

steveri replied:
I changed the parameter to simply be the full image name e.g. "stanfordaha/garnet:latest". That's more versatile and also more obvious in terms of how to use, so I think should be better in the long run...

steveri commented on mflowgen/Tile_PE/construct.py:
```diff
@@ -352,6 +352,12 @@ def construct():
+  # The registers cause short circuits that are not found until much later, after LVS.
+  # This has happened several times now, see .e.g. https://buildkite.com/tapeout-aha/fullchip/builds/358
+  synth.extend_postconditions( ["assert 'SDFQ' not in File( 'outputs/design.v' )"] )
+
```
Removed.

steveri commented:
Okay @alexcarsello I think I addressed all your concerns. Whenever you get a moment, please go back through and resolve the conversations where you can and, if things look okay, give me an approval for the final pull...thanks!

alexcarsello commented:
Looks good aside from the test failure.
